### PR TITLE
Reduce code duplication in Arrow code

### PR DESCRIPTION
**Pull request by nastra**

### CODE REVIEW

**Comment by nastra** [comment]
@rymurr it is probably easier looking at the modified files directly instead of looking at the diff when reviewing.

**Comment by nastra** [comment]
## Results on branch `master`
```
Benchmark                                                                 Mode  Cnt   Score   Error  Units
VectorizedReadFlatParquetDataBenchmark.readDatesIcebergVectorized5k         ss    5   1.451 ± 0.060   s/op
VectorizedReadFlatParquetDataBenchmark.readDatesSparkVectorized5k           ss    5   1.539 ± 0.270   s/op
VectorizedReadFlatParquetDataBenchmark.readDecimalsIcebergVectorized5k      ss    5  10.142 ± 5.595   s/op
VectorizedReadFlatParquetDataBenchmark.readDecimalsSparkVectorized5k        ss    5   6.092 ± 0.173   s/op
VectorizedReadFlatParquetDataBenchmark.readDoublesIcebergVectorized5k       ss    5   1.943 ± 0.212   s/op
VectorizedReadFlatParquetDataBenchmark.readDoublesSparkVectorized5k         ss    5   2.394 ± 0.643   s/op
VectorizedReadFlatParquetDataBenchmark.readFloatsIcebergVectorized5k        ss    5   1.897 ± 0.103   s/op
VectorizedReadFlatParquetDataBenchmark.readFloatsSparkVectorized5k          ss    5   1.921 ± 0.108   s/op
VectorizedReadFlatParquetDataBenchmark.readIntegersIcebergVectorized5k      ss    5   1.872 ± 0.119   s/op
VectorizedReadFlatParquetDataBenchmark.readIntegersSparkVectorized5k        ss    5   2.554 ± 2.118   s/op
VectorizedReadFlatParquetDataBenchmark.readLongsIcebergVectorized5k         ss    5   1.930 ± 0.141   s/op
VectorizedReadFlatParquetDataBenchmark.readLongsSparkVectorized5k           ss    5   3.741 ± 0.699   s/op
VectorizedReadFlatParquetDataBenchmark.readStringsIcebergVectorized5k       ss    5   3.458 ± 0.628   s/op
VectorizedReadFlatParquetDataBenchmark.readStringsSparkVectorized5k         ss    5   4.300 ± 4.699   s/op
VectorizedReadFlatParquetDataBenchmark.readTimestampsIcebergVectorized5k    ss    5   1.196 ± 0.146   s/op
VectorizedReadFlatParquetDataBenchmark.readTimestampsSparkVectorized5k      ss    5   1.435 ± 0.202   s/op
```

## Results on branch `arrow-refactoring`
```
Benchmark                                                                 Mode  Cnt  Score   Error  Units
VectorizedReadFlatParquetDataBenchmark.readDatesIcebergVectorized5k         ss    5  1.088 ± 0.225   s/op
VectorizedReadFlatParquetDataBenchmark.readDatesSparkVectorized5k           ss    5  1.467 ± 0.222   s/op
VectorizedReadFlatParquetDataBenchmark.readDecimalsIcebergVectorized5k      ss    5  7.077 ± 3.262   s/op
VectorizedReadFlatParquetDataBenchmark.readDecimalsSparkVectorized5k        ss    5  7.101 ± 0.722   s/op
VectorizedReadFlatParquetDataBenchmark.readDoublesIcebergVectorized5k       ss    5  1.908 ± 0.130   s/op
VectorizedReadFlatParquetDataBenchmark.readDoublesSparkVectorized5k         ss    5  1.817 ± 0.274   s/op
VectorizedReadFlatParquetDataBenchmark.readFloatsIcebergVectorized5k        ss    5  1.971 ± 0.142   s/op
VectorizedReadFlatParquetDataBenchmark.readFloatsSparkVectorized5k          ss    5  2.066 ± 0.205   s/op
VectorizedReadFlatParquetDataBenchmark.readIntegersIcebergVectorized5k      ss    5  2.197 ± 0.132   s/op
VectorizedReadFlatParquetDataBenchmark.readIntegersSparkVectorized5k        ss    5  1.797 ± 0.167   s/op
VectorizedReadFlatParquetDataBenchmark.readLongsIcebergVectorized5k         ss    5  1.954 ± 0.177   s/op
VectorizedReadFlatParquetDataBenchmark.readLongsSparkVectorized5k           ss    5  1.836 ± 0.109   s/op
VectorizedReadFlatParquetDataBenchmark.readStringsIcebergVectorized5k       ss    5  3.309 ± 0.092   s/op
VectorizedReadFlatParquetDataBenchmark.readStringsSparkVectorized5k         ss    5  3.119 ± 0.194   s/op
VectorizedReadFlatParquetDataBenchmark.readTimestampsIcebergVectorized5k    ss    5  1.156 ± 0.097   s/op
VectorizedReadFlatParquetDataBenchmark.readTimestampsSparkVectorized5k      ss    5  1.427 ± 0.135   s/op
```
## Detailed results are attached below:
[master_detailed_results.txt](https://github.com/apache/iceberg/files/6733705/master_detailed_results.txt)
[refactoring_detailed_results.txt](https://github.com/apache/iceberg/files/6733707/refactoring_detailed_results.txt)

File: arrow/src/main/java/org/apache/iceberg/arrow/vectorized/parquet/VectorizedDictionaryEncodedParquetValuesReader.java [hunk 1]
```diff
@@ -40,413 +40,183 @@ public VectorizedDictionaryEncodedParquetValuesReader(int maxDefLevel, boolean s
...
-            idx++;
-          }
-          break;
-      }
-      left -= numValues;
-      currentCount -= numValues;
+  class DictionaryIdReader extends BaseDictEncodedReader {
+    @Override
+    protected void nextVal(FieldVector vector, Dictionary dict, int idx, int currentVal, int typeWidth) {
+      ((IntVector) vector).set(idx, currentVal);
```
Comment (rymurr) [inline arrow/src/main/java/org/apache/iceberg/arrow/vectorized/parquet/VectorizedDictionaryEncodedParquetValuesReader.java:81]: Any reason this is cast to `IntVector` and the others are directly manipulating the data buffer?

**Review by rymurr — COMMENTED** [review]
One minor comment @nastra . Looks great though, the vectorised code is a lot easier to read! I love deleting code!

File: arrow/src/main/java/org/apache/iceberg/arrow/vectorized/parquet/VectorizedDictionaryEncodedParquetValuesReader.java (hunk 1 above)
Comment (nastra) [reply arrow/src/main/java/org/apache/iceberg/arrow/vectorized/parquet/VectorizedDictionaryEncodedParquetValuesReader.java:81]: Dictionary encoded vectors are always represented as `IntVector` and this is how it was done in the original code. You can't do the same thing on `FieldVector`, so we have to cast here

**Review by rymurr — APPROVED** [review]
Looks awesome @nastra

**Comment by rymurr** [comment]
@rdblue: @nastra suggested merging this w/o squashing as each commit atomically refactors a single class. WDYT?

File: arrow/src/main/java/org/apache/iceberg/arrow/vectorized/parquet/VectorizedColumnIterator.java [hunk 2]
```diff
@@ -40,7 +40,8 @@
 
   public VectorizedColumnIterator(ColumnDescriptor desc, String writerVersion, boolean setArrowValidityVector) {
     super(desc);
-    Preconditions.checkArgument(desc.getMaxRepetitionLevel() == 0,
+    Preconditions.checkArgument(
+        desc.getMaxRepetitionLevel() == 0,
```
Comment (rdblue) [inline arrow/src/main/java/org/apache/iceberg/arrow/vectorized/parquet/VectorizedColumnIterator.java:44]: Nit: this is a non-functional change that could cause commit conflicts.

**Comment by rdblue** [comment]
I'm fine either way. Whatever you'd like to do.

File: arrow/src/main/java/org/apache/iceberg/arrow/vectorized/parquet/VectorizedColumnIterator.java (hunk 2 above)
Comment (nastra) [reply arrow/src/main/java/org/apache/iceberg/arrow/vectorized/parquet/VectorizedColumnIterator.java:44]: that was the result of auto code formatting. It's the only line in this file that wasn't properly formatted, so I'd rather leave it as it is if that's ok?

**Review by rymurr — APPROVED** [review]

**Comment by rymurr** [comment]
note - rebasing to keep each atomic class refactor as a separate (revertible) commit

**Pull request merged**